x/ibc-transfer/keeper/relay.go

```diff
@@ -168,7 +196,25 @@ func (k Keeper) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, data t
 	sourcePrefix := types.GetDenomPrefix(packet.GetDestPort(), packet.GetDestChannel())
 	// NOTE: sourcePrefix contains the trailing "/"
 	prefixedDenom := sourcePrefix + data.Denom
-	voucher := sdk.NewCoin(prefixedDenom, sdk.NewIntFromUint64(data.Amount))
+
+	// construct the denomination trace from the full raw denomination
+	denomTrace := types.ParseDenomTrace(prefixedDenom)
+
+	traceHash := denomTrace.Hash()
+	if !k.HasDenomTrace(ctx, traceHash) {
+		k.SetDenomTrace(ctx, denomTrace)
+	}
+
+	voucherDenom := denomTrace.IBCDenom()
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeDenomTrace,
```

Comment:
this is... an unusual event? shouldn't a relayer just be able to calculate this from the message?

Reply:
they could calculate the hash on their own, but it's nice to map the received denom with the full path to the actually denom that is stored in the balance